daemon/src/supervisor.rs

```diff
@@ -0,0 +1,114 @@
+use std::fmt;
+
+use crate::Tasks;
+use async_trait::async_trait;
+use xtra::{Address, Context, Message};
+use xtra_productivity::xtra_productivity;
+
+/// A supervising actor reacts to messages from the actor it is supervising and restarts it based on
+/// a given policy.
+pub struct Actor<T, R> {
+    context: Context<T>,
+    ctor: Box<dyn Fn(Address<Self>) -> T + Send + 'static>,
+    tasks: Tasks,
+    restart_policy: Box<dyn Fn(R) -> bool + Send + 'static>,
+}
+
+impl<T, R> Actor<T, R>
+where
+    T: xtra::Actor,
+    R: fmt::Display + 'static,
+{
+    /// Construct a new supervisor.
+    ///
+    /// The supervisor needs to know two things:
+    /// 1. How to construct an instance of the actor.
+    /// 2. When to construct an instance of the actor.
+    pub fn new(
+        ctor: impl (Fn(Address<Self>) -> T) + Send + 'static,
+        restart_policy: impl (Fn(R) -> bool) + Send + 'static,
+    ) -> (Self, Address<T>) {
+        let (address, context) = Context::new(None);
+
+        let supervisor = Self {
+            context,
+            ctor: Box::new(ctor),
+            tasks: Tasks::default(),
+            restart_policy: Box::new(restart_policy),
+        };
+
+        (supervisor, address)
+    }
+
+    fn spawn_new(&mut self, ctx: &mut Context<Self>) {
+        tracing::info!("Spawning new instance of actor"); // TODO: Include name
+
+        let this = ctx.address().expect("we are alive");
+        let actor = (self.ctor)(this);
+
+        self.tasks.add(self.context.attach(actor));
+    }
+}
+
+// TODO: Should a supervisor act as proxy?
```

Comment:
it seems like it is already owning the actor that is being supervised, so it would be more convenient to prevent handing out any references (`xtra::Address` in this case) that could outlive the supervisor - it would be easiest to only hand out the ones to the supervisor to control the lifetime of both.

Reply:
I am currently leaning towards "no" to this question and instead hand out the address as part of the constructor like it is currently done.